File: examples/ch08/py/after.py
"""Chapter 8 · AFTER: open for extension via an injected interface.

process_payment depends on the IPaymentGateway abstraction, not on concrete
providers. Adding a provider is a new class; the core is never edited.
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Protocol, runtime_checkable


@dataclass(frozen=True)
class Receipt:
    provider: str
    amount: float
    confirmation: str


class IPaymentGateway(ABC):
    @abstractmethod
    def charge(self, amount: float) -> Receipt:
        """Charge ``amount``. Postcondition: returns a Receipt for that amount."""


class CardGateway(IPaymentGateway):
    def charge(self, amount: float) -> Receipt:
        return Receipt("card", amount, f"card-{int(amount * 100)}")


class PayPalGateway(IPaymentGateway):
    def charge(self, amount: float) -> Receipt:
        return Receipt("paypal", amount, f"pp-{int(amount * 100)}")


def process_payment(gateway: IPaymentGateway, amount: float) -> Receipt:
    """Open for extension: a new gateway is a new class, not an edit here."""
    return gateway.charge(amount)


class BadGateway(IPaymentGateway):
    """A Liskov violation: substitutable in name only — breaks the contract."""

    def charge(self, amount: float) -> Receipt:  # type: ignore[return]
        return None  # type: ignore[return-value]  # violates the postcondition


@runtime_checkable
class SupportsCharge(Protocol):
    """Structural interface for code we don't own — matches by shape, not name."""

    def charge(self, amount: float) -> Receipt: ...


class ThirdPartyWallet:
    """A third-party gateway that does NOT inherit IPaymentGateway."""

    def charge(self, amount: float) -> Receipt:
        return Receipt("wallet", amount, f"w-{int(amount * 100)}")

File: examples/ch08/ts/after.ts
// Chapter 8 · AFTER: open for extension via an injected interface.
//
// processPayment depends on the IPaymentGateway abstraction, not on concrete
// providers. Adding a provider is a new class; the core is never edited.
//
// The cross-language lesson lives here: in Python you must CHOOSE between a
// nominal interface (an ABC you inherit) and a structural one (a Protocol that
// matches by shape). In TypeScript that choice is gone — interfaces are
// structural by default. A class satisfies IPaymentGateway whether or not it
// writes `implements`, and a foreign object with the right shape just fits.

export interface Receipt {
  readonly provider: string;
  readonly amount: number;
  readonly confirmation: string;
}

// Our target interface. We own both sides, so we write `implements` on our
// gateways below for intent and a clearer compiler error — but TS would accept
// a structural match without the keyword (see InHouseGateway).
export interface IPaymentGateway {
  // Charge `amount`. Postcondition: returns a Receipt for that amount.
  charge(amount: number): Receipt;
}

export class CardGateway implements IPaymentGateway {
  charge(amount: number): Receipt {
    return { provider: "card", amount, confirmation: `card-${Math.round(amount * 100)}` };
  }
}

export class PayPalGateway implements IPaymentGateway {
  charge(amount: number): Receipt {
    return { provider: "paypal", amount, confirmation: `pp-${Math.round(amount * 100)}` };
  }
}

// Open for extension: a new gateway is a new class, not an edit here.
export function processPayment(gateway: IPaymentGateway, amount: number): Receipt {
  return gateway.charge(amount);
}

// A Liskov violation: substitutable in name only — breaks the contract. It
// declares `implements`, so it type-checks, yet returns null at runtime. We
// cast to escape the compiler the way the Python version uses `# type: ignore`.
export class BadGateway implements IPaymentGateway {
  charge(_amount: number): Receipt {
    return null as unknown as Receipt; // violates the postcondition
  }
}

// In Python this point needs a separate `Protocol`; in TS the interface above
// is ALREADY structural, so we reuse the shape. A class that never declares
// `implements` and inherits nothing of ours still satisfies it — by shape alone.
export class InHouseGateway {
  charge(amount: number): Receipt {
    return { provider: "inhouse", amount, confirmation: `ih-${Math.round(amount * 100)}` };
  }
}

// A foreign object — a plain object literal, no class, no inheritance — also
// fits IPaymentGateway the moment it has a matching `charge`. Structural typing
// is what lets `process` accept it untouched.
export const thirdPartyWallet: IPaymentGateway = {
  charge(amount: number): Receipt {
    return { provider: "wallet", amount, confirmation: `w-${Math.round(amount * 100)}` };
  },
};
